### ISSUE TITLE
Unclear how to use the optax.dpsgd optimizer with pmap

### ISSUE THREAD

**Issue by bastings** [deepmind/optax#148]
Hi,

The `optax.dpsgd` optimizer is special in that it takes per-example gradients as input, and takes care of aggregating them.

There is currently no documentation of how this implementation is supposed to be used with multiple devices (using `pmap`), and there might be a few options to do it.

@n2cholas is this something you have thought about?

**Comment by n2cholas** [comment]
Hi @bastings! This is a good question: if you naively `pmean` after the `differentially_private_aggregate`, I believe your noise scaling will be wrong. I don't think dividing your `noise_multiplier` by the number of devices would yield the right results either (though please correct me if my math is off).

I think it makes sense for optax to support this directly via a `distributed` argument. For the update function, you would need to multiply the batch size by the number of devices here: 

https://github.com/deepmind/optax/blob/30947fbc5743adc3e997c6242fa3775834862a74/optax/_src/privacy.py#L62

Then, do a `jax.lax.psum` between these two lines:

https://github.com/deepmind/optax/blob/30947fbc5743adc3e997c6242fa3775834862a74/optax/_src/privacy.py#L73-L75

The `init_fn` is more complicated. I think users normally create optimizer state then replicate it across devices, however, they need to split the `rng_key` after replicating so each device has a different one. If users first replicated their parameters then passed the replicated parameters to `GradientTransformation.init`, we could take care of correctly splitting the RNG key in distributed mode, however, I don't think this is a common pattern. Any thoughts on this?

**Comment by bastings** [comment]
Thanks @n2cholas ! Sorry for the wait.

I think we might want to have the *same* RNG on each device right? Since then we will make the same update across devices, otherwise the parameters will diverge. 

Or, we make different updates (with different noise), and we aggregate, and we need to increase the noise multiplier to make up for the averaging (this should be fine too).

So:

Pros with idea 1: it's a correct implementation of DP-SGD, we just sum the per-device gradients before adding the noise.
Cons: might be dangerous to generate the same noise on each device (but it should work fine, unless hardware RNG?).

Pros with idea 2: it's probably safer if we have different noise on each device.
Cons: It's not the "correct" implementation of DP-SGD, rather it's more like doing micro-batches.

I think I like idea 1 more, but maybe we should put a fail-safe in place that the RNGs are the same?
What do you think?

**Comment by n2cholas** [comment]
Good point, I totally missed that fact. I agree idea 1 is the better way to go! I don't think it is dangerous to generate the same noise on each device, since the whole point of prng-keys is to be able to do this type of stuff (deterministically generate random noise).

**Comment by bastings** [comment]
There's one issue with it I figured: if someone uses a hardware RNG, then the parameters on the various devices will diverge. If that happens it should be visible in terms of performance quite quickly, but it's a concern.

**Comment by nataliaponomareva** [comment]
- For the option 1, a couple of questions, If a device gets restored and gets the same seed as before, does the sequence of random numbers restart (e.g. is it the same as was used in previous batch from this device). If yes, we are better of with different seed/different sequence

- Both option 1 and 2 are "correct" - we will add the correct noise to the aggregated gradients. We just need to scale the noise on per device accordingly. Apparently it is also the option that TF choose for their impl

Tbh i would just suggest the path of least resistance. If different rng per device is easy, let just do that

**Comment by mtthss** [comment]
It seems like the issue is resolved. Please reopen if needed.
